### PR TITLE
Calculating hashcode of IntRange by pairing function

### PR DESCRIPTION
**Pull request by upadhyayap**
https://issues.apache.org/jira/browse/GROOVY-7834?jql=project%20%3D%20GROOVY%20AND%20id%20%3D%20GROOVY-7834

### CODE REVIEW

**Comment by upadhyayap** [comment]
Now calculating hash code of IntRange by taking only upper and lower bounds into consideration. This means that two IntRanges with same lower and upper bounds have same hash code irrespective of the step. will that be a problem? should i also consider the step during calculation?

File: src/test/groovy/lang/IntRangeTest.groovy [hunk 1]
```diff
@@ -138,4 +138,15 @@ class IntRangeTest extends GroovyTestCase {
         assert bs[-1..<-7].toString() == '{0, 5}'
         assert bs[20..<-8].toString() == '{2, 3}'
     }
+
+    void testHashCode(){
```
Comment (upadhyayap) [inline src/test/groovy/lang/IntRangeTest.groovy]: best place to test hashcode is HashMap. Let know if need to check this up for more range of values.

**Comment by upadhyayap** [comment]
This is way faster than iterating through whole collection.

File: src/main/groovy/lang/IntRange.java [hunk 2]
```diff
@@ -415,4 +415,14 @@ public void step(int step, Closure closure) {
...
         return adapter.asList();
     }
+
+    @Override
+    public int hashCode(){
+        int hashCode;
+        final int from = this.getFrom();
+        final int to = this.getTo();
+
+        hashCode = new Integer(((from+to+1)*(from+to))/2).intValue()+to;
```
Comment (jwagenleitner) [inline src/main/groovy/lang/IntRange.java]: No need to create a new `Integer` here.

Comment (upadhyayap) [inline src/main/groovy/lang/IntRange.java]: Correct because hashCode is already an int.

Comment (upadhyayap) [inline src/main/groovy/lang/IntRange.java]: Changed and committed.

**Comment by jwagenleitner** [comment]
I think this would work ok.  Since step is not used in `equals(IntRange)` I don't think it needs to be considered.  Not much harm in returning the same hashCode for non-equal objects, as long as the same hashCode is returned when two IntRange's are `equal(IntRange)` it should be ok.

**Comment by paulk-asert** [comment]
Agreed that step isn't a problem. I would have expected that one of reverse or inclusive would have impacted something but looking into it, I don't think they do, so I think the change is all good. I'll have to think a bit more about test coverage but +1 from me in any case. If someone gets around to committing before I get some more time to think about what else might be required, I can always add more after the fact.

**Comment by jwagenleitner** [comment]
PR merged, thanks!

**Pull request closed without merging**